devito/core/autotuning.py

```diff
@@ -159,34 +129,38 @@ def autotune(operator, args, level, mode):
     finalize_time_bounds(stepper, at_args, args, mode)
 
     # Reset profiling data
-    assert operator.profiler.name in args
-    args[operator.profiler.name] = operator.profiler.timer.reset()
+    assert operator._profiler.name in args
+    args[operator._profiler.name] = operator._profiler.timer.reset()
+
+    # Autotuning summary
+    summary = {}
+    summary['runs'] = len(timings)
+    summary['tpr'] = timesteps  # tpr -> timesteps per run
+    summary['tuned'] = dict(best)
 
-    return args
+    return args, summary
 
 
 def init_time_bounds(stepper, at_args):
     if stepper is None:
         return
     dim = stepper.dim.root
     if stepper.direction is Backward:
-        at_args[dim.max_name] = at_args[dim.max_name]
-        at_args[dim.min_name] = at_args[dim.max_name] - options['at_squeezer']
+        at_args[dim.min_name] = at_args[dim.max_name] - options['squeezer']
         if at_args[dim.max_name] < at_args[dim.min_name]:
-            warning("AutoTuner: too few time iterations; giving up")
+            warning("too few time iterations; giving up")
             return False
     else:
-        at_args[dim.min_name] = at_args[dim.min_name]
-        at_args[dim.max_name] = at_args[dim.min_name] + options['at_squeezer']
+        at_args[dim.max_name] = at_args[dim.min_name] + options['squeezer']
         if at_args[dim.min_name] > at_args[dim.max_name]:
-            warning("AutoTuner: too few time iterations; giving up")
+            warning("too few time iterations; giving up")
             return False
 
     return stepper.extent(start=at_args[dim.min_name], finish=at_args[dim.max_name])
 
 
-def check_time_bounds(stepper, at_args, args):
-    if stepper is None:
+def check_time_bounds(stepper, at_args, args, mode):
```

Comment:
So when testing using runtime autotuning, I'm getting exceptions thrown from this function.

This is because the number of timesteps needed for the autotuning is larger than the number of timesteps I asked to run, which can easily happen in aggressive mode.  But since I have no control over or knowledge of the former, I would argue that this shouldn't cause an exception.

Perhaps the desired behaviour looks like one of the below:

 - "destructive" autotuning could rerun the same time range over and over, thus never going OOB.  I understand that this is different to 'runtime' , which (as I understand?) is kind of a 'tune while doing real work' mode, similar to what Chuck added to YASK.

or

 - it should terminate and choose the best result so far when it cannot try all the combinations

and/or

 - it should remember the results so far, and continue where it left off on the next call to apply() where autotuning is still enabled

Reply:
> "destructive" autotuning could rerun the same time range over and over, thus never going OOB

that should be the current behaviour

>  I understand that this is different to 'runtime' , which (as I understand?) is kind of a 'tune while doing real work' mode, similar to what Chuck added to YASK.

correct

> it should terminate and choose the best result so far when it cannot try all the combinations

you're right, this is what I'm gonna do. Exception --> warning , and return the best block shape up until that point

Reply:
Oh, I didn't even try preemptive because AFAICT it allocates copies of all the buffers?

Reply:
I think I finally got what you mean, sorry dumb me. I'll produce a fix later/tomorrow.

Reply:
done